vllm/engine/metrics/metrics.py

```diff
@@ -0,0 +1,174 @@
+        self.stats = []
+        self.last_reset_time = now
+
+    def get_throughput(self, now: float) -> float:
+        return float(np.sum(self.stats) / (now - self.last_reset_time))
+
+    def get_mean(self) -> float:
+        return float(np.mean(self.stats)) if len(self.stats) > 0 else 0
+
+    def get_last(self) -> Union[int, float]:
+        return self.stats[-1] if len(self.stats) > 0 else 0
+
+
+class PrometheusMetric(ABC):
+    """Log Stats to Prometheus and TrackStats"""
+
+    def __init__(self,
+                 attr: str,
+                 template: str,
+                 log_local: bool = True) -> None:
+        # Attribute to get correct attibute from Stats.
+        self.attr = attr
+        # String template with $metric for local logging.
+        self.template = template
+        # Tracked Stats for Logging Locally
+        self.tracked_stats = TrackedStats()
+        # Whether this metric should log locally.
+        self.log_local = log_local
+
+    @abstractmethod
+    # Log to Prometheus and local TrackedStats.
+    def log(self, stats: Stats, labels: Dict[str, str], now: float) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    # Compute metric from local TrackedStats.
+    def _compute_metric(self, now: float) -> Union[int, float]:
+        raise NotImplementedError
+
+    # Compute metric and convert to string.
+    def get_str(self, now: float) -> str:
+        return self.template.format(self._compute_metric(now))
+
+    # Track stat for local logging.
+    def track_stat(self, stat: Union[int, float]) -> None:
+        if self.log_local:
+            self.tracked_stats.append(stat)
+
+    # Reset tracked stats for next local logging interval.
+    def reset_stats(self, now: float) -> None:
+        self.tracked_stats.reset(now)
+
+
+class CounterMetric(PrometheusMetric):
+    """CounterMetric measures thoughput (# / time)."""
+
+    def __init__(self, counter: Counter, *args, **kwargs) -> None:
+        self.counter = counter
+        super().__init__(*args, **kwargs)
+
+    def log(self, stats: Stats, labels: Dict[str, str], now: float) -> None:
+        stat = getattr(stats, self.attr, 0)
+        self.counter.add(labels, stat)
+        self.track_stat(stat)
+
+    def _compute_metric(self, now: float) -> Union[int, float]:
+        return self.tracked_stats.get_throughput(now)
+
+
+class GaugeMetric(PrometheusMetric):
+    """GaugeMetric measures a metric at a moment in time."""
+
+    def __init__(self, gauge: Gauge, *args, **kwargs) -> None:
+        self.gauge = gauge
+        super().__init__(*args, **kwargs)
+
+    def log(self, stats: Stats, labels: Dict[str, str], now: float) -> None:
+        stat = getattr(stats, self.attr, 0)
+        self.gauge.set(labels, stat)
+        self.track_stat(stat)
+
+    def _compute_metric(self, now: float) -> Union[int, float]:
+        return self.tracked_stats.get_last()
+
+
+class HistogramMetric(PrometheusMetric):
```

Comment:
I think all these classes and inheritance makes the code more complex and hard to read. Every 5 seconds we compute some stats. Some of those we log to the logger, and all of them we log to Prometheus. The old code in the metrics package was 50 lines. This new code is 270 and it ads 3 new metrics.

Reply:
Agree that the code in the old `metrics.py` was much shorter. However that file was only responsible for calling the `aioprometheus` metrics whereas in this PR, I moved much of the functionality that is currently in `LLMEngine` out into generic classes. 
- Currently the `LLMEngine` is responsible for (a) computing statistics, (b) tracking stats over the 5s interval, and (c) calling `record_metrics` with the tracked stats.

- In this PR, the `LLMEngine` is just responsible for computing statistic.`MetricLogger` holds the individual `PrometheusMetrics` which are responsible for holding the `TrackedStats` (which are currently just lists of time/value in the current implementation sitting in `LLMEngine` attributes.

The spirit of the PR is to setup the framework to make it easier to add new Metrics or Loggers without having to write a bunch of code in `LLMEngine` beyond updating the `Stats` class.